dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/event/DataRelationsValidator.java

```diff
@@ -322,4 +332,28 @@ private Program getEnrollmentProgramFromEvent(TrackerBundle bundle, Event event)
     }
     return null;
   }
+
+  /**
+   * Given a program with registration, check whether an event in the preheat or payload refers to
+   * enrollment with an existing tracked entity. This should not happen because an enrollment
+   * without a tracked entity is not allowed, see {@link
+   * org.hisp.dhis.tracker.imports.validation.validator.enrollment.MandatoryFieldsValidator}
+   *
+   * @param bundle to load the enrollment from the preheat or the payload if not in the database
+   * @param event the event of an enrollment
+   * @return whether the enrollment of the event has an existing tracked entity
+   */
+  private boolean enrollmentFromEventHasTrackedEntity(TrackerBundle bundle, Event event) {
+    Enrollment enrollment = bundle.getPreheat().getEnrollment(event.getEnrollment());
+
+    if (enrollment == null) {
+      return !Objects.isNull(
+          bundle
+              .findEnrollmentByUid(event.getEnrollment())
+              .map(org.hisp.dhis.tracker.imports.domain.Enrollment::getTrackedEntity)
+              .orElse(null));
+    }
+
+    return !Objects.isNull(enrollment.getTrackedEntity());
```

Comment:
just out of curiosity, why do you prefer `Objects.isNull` rather than` == null`?

Reply:
not a particular reason, it looks more compact to me. Also, in the previous if using the `==null` made the code difficult to read